### ISSUE TITLE
loss does not converge

### ISSUE THREAD

**Issue by DLC-jjj** [xavysp/DexiNed#125]
Hi, i have problem in training process of Pytorch version.I made no changes to the project and used the original BIPEDv2 dataset for training, and the parameters used the default parameters of the project. After training for 17 epochs, the loss barely changes. In the end, it can predict the image, but the effect is not very good. What could be the reason? Looking forward to your reply.
![1667659366500](https://user-images.githubusercontent.com/62531017/200125363-f9040337-dc54-4d82-b327-557fafade6ba.png)
![lena_std](https://user-images.githubusercontent.com/62531017/200125380-4127f343-212a-4ffa-baa3-aa217c8bc693.png)

**Comment by xavysp** [comment]
Hi can I see the tensorboad graph?

**Comment by LvGuangzu** [comment]
Have you solved this problem? I also have this problem, looking forward to your reply, if it is not solved we can communicate a bit.

**Comment by DLC-jjj** [comment]
> Have you solved this problem? I also have this problem, looking forward to your reply, if it is not solved we can communicate a bit.
I'm very sorry that I haven't solved it yet. I asked one of my classmates to try this project and encountered the same problem, which has not been solved yet.

**Comment by DLC-jjj** [comment]
> Have you solved this problem? I also have this problem, looking forward to your reply, if it is not solved we can communicate a bit.

I'm very sorry that I haven't solved it yet. I asked one of my classmates to try this project and encountered the same problem, which has not been solved yet.

**Comment by DLC-jjj** [comment]
> Hi can I see the tensorboad graph?

Very sorry I didn't have tensorboard installed. The loss fluctuates from 1.5 to 3.5 from the first epoch to the end, so there is nothing wrong with the loss curve. I asked one of my classmates to try this project and encountered the same problem, and the following person also encountered the same problem. I'm a little troubled.

**Comment by xavysp** [comment]
Well if you used dexined without changing the tensorboad part, maybe you have the data, you just need to see the graph. We need to see if there  is an improvement by epochs. The labels In edge detection  are very sensitive, in some training samples may be detected more edges than in the GT and you'll find diferen loss value  but that does not mean dexined is not training. You should check the average loss of each epoch. It happen in DL based edge detectors.

**Comment by xavysp** [comment]
> > Hi can I see the tensorboad graph?
> 
> Very sorry I didn't have tensorboard installed. The loss fluctuates from 1.5 to 3.5 from the first epoch to the end, so there is nothing wrong with the loss curve. I asked one of my classmates to try this project and encountered the same problem, and the following person also encountered the same problem. I'm a little troubled.

How many training data do you have?

**Comment by LvGuangzu** [comment]
I used the data set BIPED in the project in the training process, and the training set contains 200 pictures. My loss dropped to 0.9 in the 34th round, but it rose to 2.1 in the 35th round, and the overall loss was around 2.

**Comment by LvGuangzu** [comment]
![8a048c930d2f7e963a480bdc6c5ca31](https://user-images.githubusercontent.com/55423177/200490486-36951416-441a-4469-9615-10867f9476a1.png)
Then I trained it again using the 0.9 pth file as my pretrained model and found that he started to oscillate around 1.7.

**Comment by LvGuangzu** [comment]
I read that your experiment in the paper was performed 150K times. I don't know if it is because I trained too few times. I trained for 100 rounds and found that there was no improvement.

**Comment by DLC-jjj** [comment]
> > > 嗨，我可以看到张量板图吗？
> > 
> > 
> > 非常抱歉我没有张张。从第一个纪元结束到的，安装损失在 1.5 到 5 之间的波动量板，所以让项目失败没有任何问题。下面的人也遇到了同样的问题。我有点烦恼。
> 
> 你有多少训练数据？

Thank you for your reply, I am using BIPEDv2 data, 200 training images. The training loss is similar to the loss function curve of the previous layer reply, and there is no tendency to converge.

**Comment by xavysp** [comment]
> I used the data set BIPED in the project in the training process, and the training set contains 200 pictures. My loss dropped to 0.9 in the 34th round, but it rose to 2.1 in the 35th round, and the overall loss was around 2.

Did you change some hyperparameters? May you check with my lightweight model?
LDC: Lightweight Dense CNN for Edge Detection
I cannot find any error, just to make sure that the problem is not the data.

**Comment by LvGuangzu** [comment]
> 
Yes, I changed some hyperparameters. I didn't change any hyperparameters on the first run, but found that the loss did not converge, I think it was a problem that the learning rate dropped too quickly, and then I modified the hyperparameters. First, is_testing=False; Then I modified the learning rate to drop 10x every 20 rounds.

**Comment by LvGuangzu** [comment]
> Ok, I'll try to reproduce the LDC code over the next two days, and if it's not difficult to reproduce, I'll get back to you soon.

**Comment by LvGuangzu** [comment]
> > I used the data set BIPED in the project in the training process, and the training set contains 200 pictures. My loss dropped to 0.9 in the 34th round, but it rose to 2.1 in the 35th round, and the overall loss was around 2.
> 
> Did you change some hyperparameters? May you check with my lightweight model? LDC: Lightweight Dense CNN for Edge Detection I cannot find any error, just to make sure that the problem is not the data.

Hello, I have reproduced the LCD model, and I have modified some of the hyperparameters. Below I will list all the hyperparameters I have modified.
1. is_testing=True -> is_testing=False
2. epochs = 25 -> epochs = 50
3. adjust_lr = [6,12,18] -> adjust_lr = [12,24,36]
4. The BIPED dataset is still used for training, and resume=False is set, and the training is started from scratch.
The final result shows signs of convergence.Below is the loss curve I got after training for 50 epochs.
![pic1](https://user-images.githubusercontent.com/55423177/200608571-fa111dba-92a6-410a-959e-cbeec0938219.png)

**Comment by xavysp** [comment]
> > > I used the data set BIPED in the project in the training process, and the training set contains 200 pictures. My loss dropped to 0.9 in the 34th round, but it rose to 2.1 in the 35th round, and the overall loss was around 2.
> > 
> > 
> > Did you change some hyperparameters? May you check with my lightweight model? LDC: Lightweight Dense CNN for Edge Detection I cannot find any error, just to make sure that the problem is not the data.
> 
> Hello, I have reproduced the LCD model, and I have modified some of the hyperparameters. Below I will list all the hyperparameters I have modified.
> 
> 1. is_testing=True -> is_testing=False
> 2. epochs = 25 -> epochs = 50
> 3. adjust_lr = [6,12,18] -> adjust_lr = [12,24,36]
> 4. The BIPED dataset is still used for training, and resume=False is set, and the training is started from scratch.
>    The final result shows signs of convergence.Below is the loss curve I got after training for 50 epochs.
>    ![pic1](https://user-images.githubusercontent.com/55423177/200608571-fa111dba-92a6-410a-959e-cbeec0938219.png)

Can a see the results? For example the edge-map of Lenna image.

**Comment by xavysp** [comment]
> > > I used the data set BIPED in the project in the training process, and the training set contains 200 pictures. My loss dropped to 0.9 in the 34th round, but it rose to 2.1 in the 35th round, and the overall loss was around 2.
> > 
> > 
> > Did you change some hyperparameters? May you check with my lightweight model? LDC: Lightweight Dense CNN for Edge Detection I cannot find any error, just to make sure that the problem is not the data.
> 
> There is one place I don't understand, as shown in the figure below, what is the role of the seed here? Will it affect the results if I delete it? Maybe I was careless and didn't see the introduction of the relevant content, I hope you can help me answer it. ![I} UVFO)YREHX~{W67@9U91](https://user-images.githubusercontent.com/55423177/200609073-bd87dedf-c343-4ec3-92a7-3a3221e3f486.png)

It does not matter in a large scale, just trying to generalize the edge detection by changing seed

**Comment by LvGuangzu** [comment]
> > > > I used the data set BIPED in the project in the training process, and the training set contains 200 pictures. My loss dropped to 0.9 in the 34th round, but it rose to 2.1 in the 35th round, and the overall loss was around 2.
> > > 
> > > 
> > > Did you change some hyperparameters? May you check with my lightweight model? LDC: Lightweight Dense CNN for Edge Detection I cannot find any error, just to make sure that the problem is not the data.
> > 
> > 
> > Hello, I have reproduced the LCD model, and I have modified some of the hyperparameters. Below I will list all the hyperparameters I have modified.
> > 
> > 1. is_testing=True -> is_testing=False
> > 2. epochs = 25 -> epochs = 50
> > 3. adjust_lr = [6,12,18] -> adjust_lr = [12,24,36]
> > 4. The BIPED dataset is still used for training, and resume=False is set, and the training is started from scratch.
> >    The final result shows signs of convergence.Below is the loss curve I got after training for 50 epochs.
> >    ![pic1](https://user-images.githubusercontent.com/55423177/200608571-fa111dba-92a6-410a-959e-cbeec0938219.png)
> 
> Can a see the results? For example the edge-map of Lenna image.
Of course, I use the pth file output from the 26th round, because his loss=3.56, which is the lowest within 50 rounds.The first is the avg picture, the second is the fuse picture.
![avg_lena](https://user-images.githubusercontent.com/55423177/200716128-1573f79f-7381-4244-96e8-fe5e13f334f2.png)
![fuse_lena](https://user-images.githubusercontent.com/55423177/200716161-dfcead5e-a498-4567-9bb4-9291ea7af0bc.png)

**Comment by LvGuangzu** [comment]
> > > > I used the data set BIPED in the project in the training process, and the training set contains 200 pictures. My loss dropped to 0.9 in the 34th round, but it rose to 2.1 in the 35th round, and the overall loss was around 2.
> > > 
> > > 
> > > Did you change some hyperparameters? May you check with my lightweight model? LDC: Lightweight Dense CNN for Edge Detection I cannot find any error, just to make sure that the problem is not the data.
> > 
> > 
> > There is one place I don't understand, as shown in the figure below, what is the role of the seed here? Will it affect the results if I delete it? Maybe I was careless and didn't see the introduction of the relevant content, I hope you can help me answer it. ![I} UVFO)YREHX~{W67@9U91](https://user-images.githubusercontent.com/55423177/200609073-bd87dedf-c343-4ec3-92a7-3a3221e3f486.png)
> 
> It does not matter in a large scale, just trying to generalize the edge detection by changing seed

Can I understand that the seed here is basically useless? Can I delete it later for training and testing?Or when do I need to use seeds?

**Comment by xavysp** [comment]
Yes is useless

**Comment by xavysp** [comment]
> 

Here the lenna from de fused module first the fuse one. 
![IMG-20221109-WA0018](https://user-images.githubusercontent.com/3678862/200975095-41f1c63f-f69d-4e10-856a-14e3a29c3983.jpg)
![IMG-20221109-WA0017](https://user-images.githubusercontent.com/3678862/200975100-2d2d3ece-961f-49dc-99c2-6526f22511b5.jpg)

**Comment by LvGuangzu** [comment]
> > 
> 
> Here the lenna from de fused module then the average one. Results from [LDC](https://github.com/xavysp/LDC) ![IMG-20221109-WA0018](https://user-images.githubusercontent.com/3678862/200975095-41f1c63f-f69d-4e10-856a-14e3a29c3983.jpg) ![IMG-20221109-WA0017](https://user-images.githubusercontent.com/3678862/200975100-2d2d3ece-961f-49dc-99c2-6526f22511b5.jpg)

The effect looks very good. I would like to ask what is the final convergence loss of the BIPED dataset when you used the LDC model to train it. The catloss provided in the code I use

**Comment by xavysp** [comment]
Sorry I don't have access to my former lab, and I cannot take it. But I let you know whenever a have it

**Comment by zwz-append** [comment]
> > > > 嗨，我可以看到张量板图吗？
> > > 
> > > 
> > > 非常抱歉我没有张张。从第一个纪元结束到的，安装损失在 1.5 到 5 之间的波动量板，所以让项目失败没有任何问题。下面的人也遇到了同样的问题。我有点烦恼。
> > 
> > 
> > 你有多少训练数据？
> 
> Thank you for your reply, I am using BIPEDv2 data, 200 training images. The training loss is similar to the loss function curve of the previous layer reply, and there is no tendency to converge.



> Hi, i have problem in training process of Pytorch version.I made no changes to the project and used the original BIPEDv2 dataset for training, and the parameters used the default parameters of the project. After training for 17 epochs, the loss barely changes. In the end, it can predict the image, but the effect is not very good. What could be the reason? Looking forward to your reply. ![1667659366500](https://user-images.githubusercontent.com/62531017/200125363-f9040337-dc54-4d82-b327-557fafade6ba.png) ![lena_std](https://user-images.githubusercontent.com/62531017/200125380-4127f343-212a-4ffa-baa3-aa217c8bc693.png)

Hello, I have a question about _**dataset.py**_ when i use bipedv2 dataset for training.  I have changed the _'data_dir'_ in _**main.py**_ , and _data_types= ['aug'] -> data_types= ['real']_ in  lin 322 and 330 of _**dataset.py**_. However, the system show me that _**NotADirectoryError: [WinError 267] 目录名称无效。: 'E:\\SZUer\\codes\\CVweizhu\\DexiNed-master\\dataset-lists\\BIPEDv2\\BIPED\\edges\\imgs\\train\\rgbr\\real\\RGB_001.jpg'**_. 
I have tried to test _lin 368-374_ of _**dataset.py**_ , but it can't success. 
However I can ues bipedv2 dataset for testing. I don't know how to make it.

**Comment by LvGuangzu** [comment]
> Sorry I don't have access to my former lab, and I cannot take it. But I let you know whenever a have it

OK, thank you. I used the data enhancement operation you used. The current data volume is 288 * 200 images. Our laboratory can only use a 3090 gpu. It takes an hour and a half to train the dexed model. So I want to ask how many times you converged at that time? I saw in Dexined's paper that you iterated 150k times when doing experiments. Is that 150 rounds or 150,000 rounds?

**Comment by LvGuangzu** [comment]
> Sorry I don't have access to my former lab, and I cannot take it. But I let you know whenever a have it

OK, thank you. I used the data enhancement operation you used. The current data volume is 288 * 200 images. Our laboratory can only use a 3090 gpu. It takes an hour and a half to train one epoch of the dexined model. So I want to ask how many times you converged at that time? I saw in Dexined's paper that you iterated 150k times when doing experiments. Is that 150 rounds or 150,000 rounds?
![%4_R6$W061 9EH)D7LNR%ME](https://user-images.githubusercontent.com/55423177/207044591-e312bed8-cdbe-449a-8cad-7ed11f65e02a.png)
